### PR TITLE
Expand Schema Registry support

### PR DESCRIPTION
**Pull request by tjwp**
This PR is against the salsify-master branch.

This change completes support for the Subjects resource of the Schema Registry. Responses for the additional endpoints are not cached. The `CachedSchemaRegistry` is modified to delegate to the upstream `SchemaRegistry` so that the same API is available via either registry.

The `FakeSchemaRegistry` had to also be expanded to support testing. Tests have been refactored into a shared example that is used to test both the cached and the regular `SchemaRegistry`.

Prime: @jturkel

### CODE REVIEW

File: lib/avro_turf/schema_registry.rb [hunk 1]
```diff
@@ -28,6 +28,29 @@ def register(subject, schema)
...
   end
 
+  # List all subjects
+  def subjects
+    get('/subjects')
+  end
+
+  # List all versions for a subject
+  def subject_versions(subject)
+    get("/subjects/#{subject}/versions")
```
Comment (jturkel) [inline lib/avro_turf/schema_registry.rb]: Do we have to worry about URL escaping the subject?

File: lib/avro_turf/schema_registry.rb [hunk 2]
```diff
@@ -28,6 +28,29 @@ def register(subject, schema)
...
+  def subject_version(subject, version = 'latest')
+    get("/subjects/#{subject}/versions/#{version}")
+  end
+
+  # Check if a schema exists. Returns nil if not found.
+  def check(subject, schema)
+    data = post("/subjects/#{subject}",
+                expects: [200, 404],
+                body: { schema: schema }.to_json)
+    data unless data.has_key?("error_code")
```
Comment (jturkel) [inline lib/avro_turf/schema_registry.rb]: Will an exception already have been thrown for non-404 error codes?

File: spec/fake_schema_registry_server.rb [hunk 3]
```diff
@@ -19,7 +20,58 @@ class FakeSchemaRegistryServer < Sinatra::Base
     { schema: schema }.to_json
   end
 
+  get "/subjects" do
+    SUBJECTS.keys.to_json
+  end
+
+  get "/subjects/:subject/versions" do
+    schema_ids = SUBJECTS[params[:subject]]
```
Comment (jturkel) [inline spec/fake_schema_registry_server.rb]: Should we handle a 404 scenario?

File: spec/fake_schema_registry_server.rb [hunk 4]
```diff
@@ -19,7 +20,58 @@ class FakeSchemaRegistryServer < Sinatra::Base
...
+  get "/subjects/:subject/versions/:version" do
+    schema_ids = SUBJECTS[params[:subject]]
+
+    schema_id = if params[:version] == 'latest'
+                  schema_ids.last
+                else
+                  schema_ids.at(Integer(params[:version]) - 1)
+                end
+
+    schema = SCHEMAS.at(schema_id)
```
Comment (jturkel) [inline spec/fake_schema_registry_server.rb]: Should we handle a version (or subject) that doesn't exist?

**Comment by jturkel** [comment]
A few minor questions then :shipit:

File: lib/avro_turf/schema_registry.rb (hunk 2 above)
Comment (tjwp) [inline lib/avro_turf/schema_registry.rb]: Yes, anything besides the expected codes will raise an exception.

File: lib/avro_turf/schema_registry.rb (hunk 1 above)
Comment (tjwp) [inline lib/avro_turf/schema_registry.rb]: I don't think so. As used they should only contain: [A-Za-z0-9_\.]

File: spec/fake_schema_registry_server.rb (hunk 3 above)
Comment (tjwp) [inline spec/fake_schema_registry_server.rb]: Added

**Comment by tjwp** [comment]
@jturkel Added error responses

**Comment by jturkel** [comment]
:+1:

**Comment by tjwp** [comment]
Merged to salsify-master.

**Pull request closed without merging**